### ISSUE TITLE
Question: Pausing and restarting a game

### ISSUE THREAD

**Issue by ajchang84** [bberak/nature-of-code-app#3]
Greetings. I am a novice at game design and have two questions. I have a Modal component that exists outside of the GameEngine. It acts as a menu when you pause the game and is shown when this.state.pause is true

For pausing, I save it as this.state.pause on my React component containing the GameEngine which can be read by my gameSettings entity:
      gameSettings: { getState: ()=>this.state }
This way, in my systems I can check if entities['gameSettings'].getState().pause is false before doing any further game logic.

1) Is this the correct way of pausing a game or is there suppose to be some way to just stop the engine from updating somehow?

2) If this same menu has a restart game button, what's the best way to restart the game to the beginning? Am I suppose to have a system that just resets the entities to initial or is there some way to reboot the GameEngine to the beginning?

I guess the crux of my problem is that I feel like a Game Menu shouldn't be considered an entity within the GameEngine, hence trying to make menu navigations exist at the react native level yet interact with the GameEngine somehow.

Great library and thanks for any feedback.

**Comment by bberak** [comment]
Hi @ajchang84,

Thanks for getting in touch. Your intuition about leaving the Game Menu out of your entities is spot on (although I like your workaround for pausing and starting the game).

1. The GameEngine component has two methods (undocumented unfortunately) for starting and stopping:

```
	pauseClickHandler = () => {
		this.refs.gameEngine.stop();
	}

	resumeClickHandler = () => {
		this.refs.gameEngine.start();
	}

	render(){
		<GameEngine ref={"gameEngine"} entities={INITIAL_GAME_ENTITIES} />
	}
	
```

2. In terms of restarting the game back to it's initial state - I've just encountered this use case myself and have realized that RNGE doesn't handle this elegantly at all. Currently, the only way to achieve this would be to force the GameEngine component to unmount and remount somehow, which is really hacky imho. I've got a couple of ideas here:

```
	this.gameEngine.restart(); //-- Resets the game engine's internal state based on the original props

	this.gameEngine.start(NEW_ENTITIES) //-- Sets the game engine's internal state based on the passed in arguments. Could be handy for level switches etc

	<GameEngine entities={this.state.currentLevel} /> //-- Game engine will detect changes in the entities prop and will force a reset. Also sits well with React's declarative paradigm

```

I think a combination of the above should give game designers enough flexibility to have restarts / save games / level switches etc.. Let me know your thoughts. This should take me a day or two to implement and push..

I'm always looking for feedback so please let me know how you progress with your game and if you find any sticky points etc.

Cheers!

**Comment by ajchang84** [comment]
Hey @bberak,

Oh yes, those GameEngine methods for pausing the engine will do nicely!

All the restart options sound great; the first two looks to be the most intuitive though I wonder if it's better to keep restart as the method name for handling restarts with NEW_ENTITIES.

The third option definitely will make it more React-like. I've tried putting state in entity before and wondered why it wouldn't re-render the GameEngine.

Our game is coming along nicely; we've solved some snags in touch controls (the player moves one tile at a time through swipe gestures). Ended up with using 'move' to determine a projected direction and 'end' to fire the actual position change. 

Then there is the issue of game time. If we want enemies to spawn along the edges of the screen at a certain interval, what is the best way to implement this? For now, I have a spawn system that adds 16 to our global time variable per tick, and once it reaches a spawn time threshold it spawns the enemy and resets time back to 0. Should I be using the time delta prop?

Thanks

Just hit another snag, is there a proper way to stop the engine from within systems? For example, I have a Matter.Events.on collision in a system that should pause the game from further updates after colliding with the enemy. Perhaps the events shouldn't reside within systems and be placed outside of GameEngine component?

**Comment by bberak** [comment]
Glad to hear your game is coming along @ajchang84 :) Thanks for your input on the start/stop/restart feature - I'll keep you posted on any changes.

If your enemy system needs to spawn enemies at regular intervals - then counting ticks is the simplest way to go. The delta prop would be more accurate than adding 16 every tick since it gives you the exact time since the last tick (could be more or less than 16ms) - but we're talking millis here, so it probably won't make much difference. You could also store a ```timeSinceLastEnemy``` variable and diff that with ```time.current``` in your enemy system to determine if the threshold has been reached.

Alternatively, if your enemy spawns are relatively sporadic, you can use some sort of random or noise function to determine whether to spawn or not. This of course is simple, but quite crude..:

```
let test = rand(0, 1);

if (test < 0.01) {
	//-- spawn enemy
}
```

As for stopping the game within the systems, you can use the events + dispatch args to achieve this. This is useful for decoupling external components from the GameEngine whilst still allowing for two-way communication:

```
const collisionSystem = (entities, { dispatch }) => {
	
	//-- Improvising MatterJS code here
	Matter.Events.on("collision", () => { 
	
		//-- Somethings collided
		dispatch({ type: "game-over" })
	})

	return entities;
}

...

eventHandler = (e) => {
	if (e.type === "game-over") {
		this.refs.gameEngine.stop();
		this.showMenu();
	}
}

<GameEngine ref={"gameEngine"} onEvent={eventHandler} entities{...} />
```

You can also dispatch events into your systems externally:

```

this.refs.gameEngine.dispatch({ type: "difficulty", level: "novice" });

...

const enemySystem = (entities, { events }) => {
	let difficulty = events.find(e => e.type === "difficulty");

	if (difficulty && difficulty.level === "novice") {
		//-- Make the enemies dumber from here on in..
	}

	return entities;
}
```

Lastly, since you mentioned that you've been working through some issues with the touch controls, I should mention that the default touch processing logic can be overriden to suit whatever your needs may be (although it sounds like you've solved your problems in any case).

```
const specializedTouchProcessor = touches => {
	return {
		process(type, payload) {
			//-- Process get's called when the raw React Native touch events are detected.
			//-- Type can be "start" || "move" || "end".
			//-- Once you have detected a particular movement or gesture, push it into the touches array. This
			//-- is the same array that gets passed into your systems

			touches.push({ type: "circular", radius: 30, speed: 2 })
		},
		end() {
			//-- End will be called when the GameEngine unmounts
			//-- You can use this method to do cleanup or dispose of any listeners or streams 
			//-- you may have created to help you detect your domain-specific gestures.
		}
	}
}

...

<GameEngine touchProcessor={specializedTouchProcessor} entities={...} />

...

const movementSystem = (entities, { touches }) => {
	let circular = touches.find(x => x.type === "circular") {
		//-- The user made a circular motion with his finger
	}

	return entities;
}
```

Let me know if that helps, cheers!

**Comment by ajchang84** [comment]
Yes, this is all valuable feedback. Thank you so much for taking the time. The dispatch/events args really illuminate the flexibility of this library. Will keep in touch!

**Comment by bberak** [comment]
Hi @ajchang84,

Just letting you know that I've published a new version of RNGE (0.8.9). With this update you change changed the state of the GameEngine (the entities) like so:

```
this.refs.gameEngine.swap(NEW_ENTITIES);
```

The ```swap``` method will allow you to reset your game to the beginning or move to a  new level. If resetting, you might need to take  a copy/clone of the initial state, otherwise you might not be actually resetting the state at all. Let me know if there are any issues.

Also, I opted to hold off on the declarative style entity updates because it caused some valid but unexpected behaviour which might surprise existing RNGE users. I'll add support for this declarative style update in an upcoming release where I'll bump the major or minor version.

That said, you can now replace the start/stop calls with a new ```running``` prop:

```
<GameEngine entities={...} running={!this.state.menuVisible} />
```

Cheers!

**Comment by ajchang84** [comment]
Hey @bberak, 

Thanks! I have restarting the game now working perfectly fine with the new swap method. I'll let you know of any further feedback.
